python/ray/train/utils.py

```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
     """
     unfinished = remote_values.copy()
     dead_worker_indexes = []  # Store the indexes of the failed workers.
+
+    at_least_one_failed_worker = False
     while len(unfinished) > 0:
-        finished, unfinished = ray.wait(unfinished)
+        # Once a worker has failed, we add a timeout for getting the
+        # results for the remaining workers.
+        # This is to avoid situations where the remaining workers
+        # are alive, but hanging on collective calls because other
+        # workers have failed.
+        timeout = (
+            REMAINING_WORKERS_GRACE_PERIOD_S if at_least_one_failed_worker else None
+        )
+        finished, unfinished = ray.wait(unfinished, timeout=timeout)
+
+        if at_least_one_failed_worker and len(unfinished) > 0:
+            # If at least one worker has failed, but there are still workers
+            # that we are waiting on results from, these workers are hanging.
+            # Treat these workers as dead workers as well.
```

Comment:
Do we actually need to treat these as dead? Or is it sufficient that we break out of this `ray.wait` call and only return the failed worker as dead?

Reply:
That's a good point, we don't need to treat them as dead. Will make the change.

Reply:
Actually no, these have to be marked as dead because new actors may not be created. And these actors are no longer usable since they are hanging on a particular method execution.